File: src/matrices/jared/LRA.jl
#using Pkg; Pkg.add("Arpack");
using LinearAlgebra
using SparseArrays
#using Arpack: eigs
using ArnoldiMethod


offset(A) = A .+ 1e-5im

function truncate_eigens(eigen_vals, eigen_vecs, cutoff)
    for i in eachindex(eigen_vals)
        # complex magnitude abs()
        if abs(eigen_vals[i]) > cutoff
            eigen_vecs[:, i] .= 0
            eigen_vals[i] = 0
        end
    end
    return eigen_vals, eigen_vecs
end

# takes nxn size
# returns ComplexF64 laplace operator
function construct_laplace(n)
    #Construct a sparse diagonal matrix from Pairs of vectors and diagonals.
    #Each vector (second) will be placed on the (first) diagonal.
    return spdiagm(-1 => offset(ones(ComplexF64, n-1)),
                    0 => offset(-2*ones(ComplexF64,n)),
                    1 => offset(ones(ComplexF64,n-1)))
end

# takes nxn size
# returns ComplexF64 position operator
function construct_pos(n)
    return spdiagm(-1 => (sqrt.(1:n-1)),
                    1 => (sqrt.(1:n-1)))
end

module eigen_decompose
    using LinearAlgebra
    using SparseArrays
    using Arpack: eigs
    #using ArnoldiMethod

    export EigenDecomp

    struct EigenDecomp #<: AbstractArray{XXX, 2}
        #F :: Eigen 
        e_vals :: ComplexF64
        e_vecs :: Array
        function EigenDecomp(A)#(M:Array{XXX, 2})
            #decomp, history = partialschur(A, nev=10, tol=1e-6, which=SR());
            #e_vals, e_vecs = partialeigen(decomp)
            e_vals, e_vecs = eigs(A, nev=10)
            #e_vecs = eigvecs(A)
        end
    end

    function Base.show(io::IO, M::EigenDecomp)
        # type inference cannot occur here
        print(io,M.e_vals,"\n")
        print(io,M.e_vecs,"\n")
    end
end

n = 14
H = construct_laplace(n)

test_struct = eigen_decompose.EigenDecomp(Matrix(H))

"
#display(H)
# currently converting from sparse 
E = eigvals(Tridiagonal(H))
ϕ = eigvecs(Tridiagonal(H))
#eigs(H, nev = 2, which=:SM)
#display(ϕ)
E, ϕ = truncate_eigens(E, ϕ, 3)
S = deepcopy(ϕ)
X = construct_pos(n)
Sꜛ = conj(transpose(S))
Xₛ = Sꜛ * X * S
xₛ = eigvals(Tridiagonal(Xₛ))
ψ  = eigvecs(Tridiagonal(Xₛ))
#Iₙ = I(n)
E_iη = offset(E)
Gᵣ = inv( Matrix(Diagonal(E_iη) - H) )
"

File: src/matrices/jared/LRA.py
from scipy.sparse import diags
import numpy as np
from scipy.sparse.linalg import eigs

offset = lambda arr: np.complex64(arr) + complex(0,1e-5)

def truncate_eigens(eigen_vals, eigen_vecs, cutoff):
    for i,e in enumerate(eigen_vals):
        if ( cmplx_mag(e) > cutoff ):
            eigen_vecs[i] = np.zeros( len(np.transpose(eigen_vecs)) )
            # remove eigen energy too?
            eigen_vals[i] = 0
    return eigen_vals, eigen_vecs

def construct_laplace(n):
    return (np.diag(offset(np.ones(n-1)) , -1) +
            np.diag(offset(np.full(n,-2)))     +
            np.diag(offset(np.ones(n-1)), 1) )

def construct_pos(n):
    return (np.diag( [np.sqrt(complex(i+1,0)) for i in range(n-1)], -1) +
            np.diag( [np.sqrt(complex(i+1,0)) for i in range(n-1)], 1) )

def cmplx_mag(u):
    return np.sqrt(u * np.conjugate(u))

n = 4
H = construct_laplace(n)
E, phi = truncate_eigens(*eigs(H), 3)
S = phi # matrix of eigen vectors as columns
X = construct_pos(n)
S_dagger = np.conjugate( np.transpose(S) )
Xs = S_dagger * X * S
xs , psi = eigs(Xs)

I = np.identity(n)
E_ieta = offset(E)
G_R = (E_ieta) * I - H # minus self energy terms
